bazel/re2c.bzl

```diff
@@ -0,0 +1,81 @@
+# re2c generator rules - assuming that re2c is available as '@re2c//:re2c'.
+
+visibility("public")
+
+def _re2c_impl(ctx):
+    srcs = ctx.files.srcs
+    if len(srcs) != 1:
+        fail("Attribute `srcs` must contain exactly 1 '.re2c' file.")
+    outs = ctx.outputs.outs
+    if not outs or [o for o in outs if o.path.split(".")[-1] not in ["c", "cc", "h"]]:
+        fail("Attribute `outs` must contain at least 1 '.cc' or '.h' file.")
+    err_file = ctx.actions.declare_file(ctx.attr.name + ".err")
+    ctx.actions.run_shell(
+        outputs = outs + [err_file],
+        tools = [ctx.executable._re2c_tool],
+        inputs = ctx.files.srcs + ctx.files.data,
+        command = """
+            line_args=""
+            [ -z "${re2c_parse_first_line}" ] && \\
+                line_args="$(sed -rne "1s,// re2c .INPUT -o .OUTPUT (.+),\\1,p" "{src}")"
+            "{re2c}" "{src}" -o "{out}" ${{line_args}} {re2c_args} {includes} > "{err}" 2>&1
+            cat "{err}" >> "{out}"
+        """.format(
+            re2c = ctx.executable._re2c_tool.path,
+            src = srcs[0].path,
+            out = outs[0].path,
+            err = err_file.path,
+            includes = " ".join(list(dict([("-I" + f.dirname, 1) for f in ctx.files.data]))),
+            re2c_parse_first_line = "1" if ctx.attr.re2c_parse_first_line else "",
+            re2c_args = " ".join(ctx.attr.re2c_args),
+        ),
+        mnemonic = "RE2C",
+        progress_message = "Generating re2c %{label}",
+    )
+
+_re2c_attr = {
+    "srcs": attr.label_list(
+        doc = "The re2c source.",
+        allow_files = [
+            ".re",
+            ".re2c",
+        ],
+        mandatory = True,
+    ),
+    "data": attr.label_list(
+        doc = "Other input data files (e.g. include files).",
+        allow_files = True,
+        mandatory = False,
+    ),
+    "outs": attr.output_list(
+        doc = "The generated output source.",
+        mandatory = True,
+    ),
+    "re2c_args": attr.string_list(
+        doc = """
+              Arguments to re2c invocation. These will be applied after arguments
+              parsed from source if `re2c_parse_first_line` is active.
+              """,
+    ),
+    "re2c_parse_first_line": attr.bool(
+        doc = """
+              Whether parse the first line of the re2c input for arguments.
+              For this to work the first line must start with:
+                `// re2c $INPUT -o $OUTPUT`.
+              """,
+    ),
+    "_re2c_tool": attr.label(
+        doc = "The target of the 're2c' executable.",
+        default = Label("//:re2c"),
+        allow_single_file = True,
+        executable = True,
+        cfg = "exec",
+    ),
+}
+
+re2c_gen = rule(
+    attrs = _re2c_attr,
+    doc = """Use re2c tool to create an output file (e.g. '.cc') from a '.re2' input.""",
```

Comment:
The usual file extension is .re, not .re2.

Reply:
Done.